B 20619 prime create mobile home int

KonstanceH opened the pull request:
## [B-20619](https://www13.v1host.com/USTRANSCOM38/story.mvc/Summary?oidToken=Story%3A982505)

## Summary

This PR adds the capability for Prime counselors to create mobile home shipments for existing moves.

## Verification Steps for Reviewers

These are to be checked by a reviewer.

- [ ] Has the branch been pulled in and checked out?
- [ ] Have the BL acceptance criteria been met for this change?
- [ ] Was the CircleCI build successful?
- [ ] Has the code been reviewed from a standards and best practices point of view?

### How to test

1. Login as Prime
2. Choose any move
3. Click "Create Shipment" button and choose "MOBILE_HOME"
4. Confirm unique mobile home shipment fields show
5. Confirm to see mobile home shipment success banner 
6. From the confirmation banner you can use that id for this query to check in DB that it was added
`select id, shipment_type, status  from mto_shipments ms where id ='shipment id'`

### Frontend

- [ ] There are no aXe warnings for UI.
- [ ] This works in [Supported Browsers and their phone views](https://transcom.github.io/mymove-docs/docs/adrs/Browser-Support/#minimum-browser-requirements) (Chrome, Firefox, Edge).
- [ ] There are no new console errors in the browser devtools.
- [ ] There are no new console errors in the test output.
- [ ] If this PR adds a new component to Storybook, it ensures the component is fully responsive, OR if it is intentionally not, a wrapping div using the `officeApp` class or custom `min-width` styling is used to hide any states the would not be visible to the user.
- [ ] This change meets the standards for [Section 508 compliance](https://www.ssa.gov/accessibility/andi/help/install.html).

### Backend

- [ ] Code follows the guidelines for [Logging](https://transcom.github.io/mymove-docs/docs/getting-started/development/logging).
- [ ] The requirements listed in [Querying the Database Safely](https://transcom.github.io/mymove-docs/docs/backend/guides/golang-guide#querying-the-database-safely) have been satisfied.

## Screenshots
**_Mobile Home Form_**
<img width="1072" alt="Screenshot 2024-09-13 at 1 53 53 PM" src="https://github.com/user-attachments/assets/c01153ae-da82-4237-b078-22e4ac39b6a2">

**_Mobile Home added to shipment types dropdown list_**
<img width="910" alt="Screenshot 2024-09-13 at 1 53 25 PM" src="https://github.com/user-attachments/assets/91db7211-6582-4f0f-805e-82f58b845316">

**_Created Mobile Home Shipment Successful banner_**
<img width="1141" alt="Screenshot 2024-09-13 at 1 51 05 PM" src="https://github.com/user-attachments/assets/3b726358-ac82-4654-850a-cbce76a6c2ff">

**_Mobile Home Shipment specific fields_**
<img width="953" alt="Screenshot 2024-09-17 at 2 10 49 PM" src="https://github.com/user-attachments/assets/862ca8a6-779d-44c9-999f-4a287cad18e9">

robot-mymove commented:
<!--
  0 failure: 
  1 warning:  Files located in ...
  
  
  DangerID: danger-id-Danger;
-->


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  Files located in legacy directories (src/shared or src/scenes) have
been edited. Are you sure you don’t want to also relocate them to the new [file structure](https://transcom.github.io/mymove-docs/docs/dev/contributing/frontend/frontend#file-layout--naming)?

View the [frontend file org ADR](https://github.com/transcom/mymove/blob/main/docs/adr/0048-frontend-file-org.md) for more information
  </td>
    </tr>
  </tbody>
</table>



<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 9afc29a49737388334fe42209bdbd549bd96ca7e
</p>

taeJungCaci commented:
I see this section for Boat shipment. Should this also be the same for mobile homes? 
![image](https://github.com/user-attachments/assets/105a6cf9-c6ab-4160-97b0-b25ab83c4140)

JamesHawks224 commented on pkg/handlers/primeapiv2/mto_shipment.go:
```diff
@@ -57,6 +57,23 @@ func (h CreateMTOShipmentHandler) Handle(params mtoshipmentops.CreateMTOShipment
 			}
 
+			/** Feature Flag - Mobile Home Shipment **/
+			featureFlagMobileHome := "mobile_home"
```
"mobile_home"
This string literal occurs in multiple places.  Should it be made a constant and referenced in the code?

KonstanceH replied:
I was gonna just make the featureflagname a global string array and iterate through it to set the variables the same, but the work vs payout of that is just not worth it to me. And I think this is outside of the scope of this story. I know we are already in talks about reconstructing how to approach feature flags so I think the effort for this would have to be added to an issue or future innovation task.

KonstanceH replied:
ah okay I completely misunderstood you the first time.

JamesHawks224 commented on pkg/handlers/primeapiv2/mto_shipment.go:
```diff
@@ -57,6 +57,23 @@ func (h CreateMTOShipmentHandler) Handle(params mtoshipmentops.CreateMTOShipment
+			flagMH, err := h.FeatureFlagFetcher().GetBooleanFlagForUser(params.HTTPRequest.Context(), appCtx, featureFlagMobileHome, map[string]string{})
+			if err != nil {
+				appCtx.Logger().Error("Error fetching feature flagMH", zap.String("featureFlagKey", featureFlagMobileHome), zap.Error(err))
+				isMobileHomeFeatureOn = false
```
isMobileHomeFeatureOn
This was already initialized to false at line 62.  Does it need to be initialized again after the nil check?

KonstanceH replied:
It's initialized false at the top and set to false later, but I think this could be initialized as false at the top and then at the end after some validation checks it can be set to true. It's just that in this moment the featureflag is off.

JamesHawks224 replied:
I'm just referring to redundant unnecessary code.  If its initialized at the top, there's not a point to setting it to false in conditional... its already false.

KonstanceH replied:
oh okay I see what you mean. I'll go ahead and remove it

JamesHawks224 commented on pkg/handlers/primeapiv3/mto_shipment.go:
"mobile_home"
This string literal occurs in multiple places. Should it be made a constant and referenced in the code?

isMobileHomeFeatureOn
This was already initialized to false at line 62. Does it need to be initialized again after the nil check?

KonstanceH replied:
honestly I'm not sure, I'll poke around and see what's standard or if I should/could make it a constant somewhere for the feature flags.

KonstanceH replied:
I poked around and came to the conclusion that this would be outside of the scope for this story. I left comments responding to yours above.

JamesHawks224 commented on src/pages/PrimeUI/Shipment/PrimeUIShipmentCreate.jsx:
```diff
@@ -351,6 +411,23 @@ const PrimeUIShipmentCreate = ({ setFlashMessage }) => {
+      is: (shipmentType) => shipmentType === SHIPMENT_TYPES.MOBILE_HOME,
+      then: () =>
+        Yup.object().shape({
+          year: Yup.number().positive('Must be a postive number').required('Required'),
```
year, length, width, height
Is there any range checking done for these values to ensure its valid?

KonstanceH replied:
There's a a range check on the front end check for length, width, height inches. None for the dimensions in feet for the mobile home. I asked in a different story and was told there is no validation for the dimensions of the mobile home in general like the boat shipment. There is only validation on the backend for the year so you'd have to submit the form to get any feedback about that.

KonstanceH replied:
I added a max to the year, and dimensions(in feet) just so that the user can't enter an invalid field based on its type.

JamesHawks224 reviewed (changes requested):
testing: success
code: see comments

KonstanceH commented:
> I see this section for Boat shipment. Should this also be the same for mobile homes? ![image](https://private-user-images.githubusercontent.com/146856854/367425204-105a6cf9-c6ab-4160-97b0-b25ab83c4140.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjYyNTc0MDUsIm5iZiI6MTcyNjI1NzEwNSwicGF0aCI6Ii8xNDY4NTY4NTQvMzY3NDI1MjA0LTEwNWE2Y2Y5LWM2YWItNDE2MC05N2IwLWIyNWFiODNjNDE0MC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwOTEzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDkxM1QxOTUxNDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zZGNhOGEyMWZjNmExMTY1YmJiYWE2ODlkNzcyMGE4ZjAyMDI1YmI3MzM3YzdjMjcwYzI5NjIwNWJjOTU1MDA4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.z1XChKjlN1a9rbEJUV4vDCqQs6Ft1mZIm4lpVyvmq6M)
> 
> Also, I think mobileHomeShipment needs to show in the Prime GET api

I definitely added a section for mobile home too... I must of overwrote it when I merged in main. I'll add these back in now

KonstanceH commented:
> > I see this section for Boat shipment. Should this also be the same for mobile homes? ![image](https://private-user-images.githubusercontent.com/146856854/367425204-105a6cf9-c6ab-4160-97b0-b25ab83c4140.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjYyNTc0MDUsIm5iZiI6MTcyNjI1NzEwNSwicGF0aCI6Ii8xNDY4NTY4NTQvMzY3NDI1MjA0LTEwNWE2Y2Y5LWM2YWItNDE2MC05N2IwLWIyNWFiODNjNDE0MC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwOTEzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDkxM1QxOTUxNDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zZGNhOGEyMWZjNmExMTY1YmJiYWE2ODlkNzcyMGE4ZjAyMDI1YmI3MzM3YzdjMjcwYzI5NjIwNWJjOTU1MDA4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.z1XChKjlN1a9rbEJUV4vDCqQs6Ft1mZIm4lpVyvmq6M)
> > Also, I think mobileHomeShipment needs to show in the Prime GET api
> 
> I definitely added a section for mobile home too... I must of overwrote it when I merged in main. I'll add these back in now

Okay idk how but I guess it was taken out of the model to payload so I just added it back and added a screenshot of it in the description

taeJungCaci commented on pkg/handlers/primeapiv3/payloads/model_to_payload.go:
Should these also be added in prime v1 and v2?

taeJungCaci replied:
@brooklyn-welsh

taeJungCaci replied:
maybe in pricing stories?

KonstanceH replied:
I tried adding them in and the behaviors been a little weird so I'm going to add a note to add these in to the pricing stories during planning.

taeJungCaci reviewed (approved):
LGTM
